### DISCUSSION TITLE
Any way to re-position the cursor _after_ expansion?

### DISCUSSION THREAD

**Discussion by bew** [L3MON4D3/LuaSnip#1271] [Q&A] [answered]
For Lua I'm making a simple `r` snippet for `return`, but I want the following behavior:
> Add a space after `return` only if needed, but always leave the cursor after the space.

So that expanding with (`|` is the cursor):
```
r|foo
```
Would insert `return ` (with space) and end up with `return |foo`
And:
```
r| foo
```
Would insert `return` (no space, but move cursor to the right) and end up with `return |foo`

I tried to use inject the termcodes for moving the cursor but that doesn't work / inserts garbage:
```lua
s({trig = "r", desc = "return ..."}, {
  ls.function_node(function()
    local line = vim.api.nvim_get_current_line()
    local col = vim.fn.col(".")
    local char = line:sub(col, col)
    if char == " " then
      local move_right = vim.api.nvim_replace_termcodes("<C-g>U<Right>", true, false, true)
      return "return" .. move_right
    else
      return "return "
    end
  end)
})
```
And I couldn't find a `post_expand` callback 🤔

**Comment by L3MON4D3** [comment, marked as answer]
Oh, I think I have a nice way of achieving this: instead of moving the cursor after expansion, include any whitespace _behind_ the cursor in the trigger!
While this is not possible with the trigger using the `"plain"` `trigEngine`, you can use `resolveExpandParams` to manipulate the region that is treated as the trigger (ie. deleted before expansion).

Moving the cursor reliably is more complicated I think, luasnip has a kind of queue for all movement-commands and I'd have to expose some way to enqueue into it (and add a callback that is invoked at the right time :D)

**Reply by bew** [reply]
Awesome, I managed to get the behavior I wanted with (using my dsl for snip def):
```lua
-- NOTE: Adds the space after `return` only if needed,
--   but _always_ leave the cursor after the space.
snip(
  "r",
  {
    desc = "return ...",
    -- Tweak what will be removed exactly before snip expansion
    -- It is used to remove extra spaces and always leave cursor after space after expansion.
    --
    -- SEE: https://github.com/L3MON4D3/LuaSnip/discussions/1271
    -- REF: https://github.com/L3MON4D3/LuaSnip/blob/33b06d72d220aa56/lua/luasnip/extras/postfix.lua#L13
    resolveExpandParams = function(_snip, line_to_cursor, matched, _captures)
      local pos0
      do
        local c = vim.api.nvim_win_get_cursor(0)
        pos0 = { row = c[1] - 1, col = c[2] }
      end
      local line = vim.api.nvim_get_current_line()
      local line_after_cursor = line:sub(vim.fn.col".")
      local extra_spaces = line_after_cursor:match"^%s+" or ""
      return {
        clear_region = {
          from = {pos0.row, pos0.col - #matched}, -- before snip match
          to = {pos0.row, pos0.col + #extra_spaces}, -- after spaces
        }
      }
    end,
  },
  {
    t"return "
  }
)
```

Will def' play with this and refactor it for use in more snippets :heart:
Love the flexibility of the plugin ✨✨
